_posts/2024-08-06-datafusion-python-udf-comparisons.md

```diff
@@ -0,0 +1,595 @@
+every project that is using this shared UDF will benefit from those updates.
+
+When deciding to use this approach, it’s worth considering how much you think you’ll actually
+benefit from the rust implementation to decide if it’s worth the additional effort to maintain and
+deploy the python wheels you generate. It is certainly not necessary for every use case.
+
+Due to the excellent work by the python arrow team, we can simplify our work to needing only two
+dependencies on the rust side, [arrow-rs](https://github.com/apache/arrow-rs) and
+[pyo3](https://pyo3.rs/). I have posted a [minimal example](https://github.com/timsaucer/tuple_filter_example).
+You’ll need [maturin](https://github.com/PyO3/maturin) to build the project, and I recommend using
+release mode when building.
+
+```bash
+maturin develop --release
+```
+
+When you write your UDF in rust you generally will need to take these steps
+
+1. Write a function description that takes in some number of python generic objects.
+2. Convert these objects to Arrow Arrays of the appropriate type(s).
+3. Perform your computation and create a resultant Array.
+3. Convert the array into a python generic object.
+
+For the conversion to and from python objects, we can take advantage of the
+`ArrayData::from_pyarrow_bound` and `ArrayData::to_pyarrow` functions.  All that remains is to
+perform your computation.
+
+We are going to demonstrate doing this computation in two ways. The first is to mimic what we’ve
+done in the above approach using pyarrow. In the second we demonstrate iterating through the three
+arrays ourselves.
+
+In our first approach, we can expect the performance to be nearly identical to when we used the
+pyarrow compute functions. On the rust side we will have slightly less overhead but the heavy
+lifting portions of the code are essentially the same between this rust implementation and the
+pyarrow approach above.
+
+The reason for demonstrating this, even though it doesn’t provide a significant speedup over
+python, is to primarily demonstrate how to make the python to rust with python wrapper
+transition. In the second implementation you can see how we can iterate through all of the arrays
+ourselves.
+
+```rust
+#[pyfunction]
+pub fn tuple_filter_fn(
+    py: Python<'_>,
+    partkey_expr: &Bound<'_, PyAny>,
+    suppkey_expr: &Bound<'_, PyAny>,
+    returnflag_expr: &Bound<'_, PyAny>,
+) -> PyResult<Py<PyAny>> {
+    let partkey_arr: PrimitiveArray<Int64Type> =
+        ArrayData::from_pyarrow_bound(partkey_expr)?.into();
+    let suppkey_arr: PrimitiveArray<Int64Type> =
+        ArrayData::from_pyarrow_bound(suppkey_expr)?.into();
+    let returnflag_arr: StringArray = ArrayData::from_pyarrow_bound(returnflag_expr)?.into();
+
+    let values_of_interest = vec![
+        (1530, 4031, "N".to_string()),
+        (6530, 1531, "N".to_string()),
+        (5618, 619, "N".to_string()),
+        (8118, 8119, "N".to_string()),
+    ];
```

Comment:
Ah one big difference here is that you're hard-coding the values of interest in Rust at compile time rather than the user being able to supply it at runtime. There's probably in theory some ways around this, where you can export to rust a wrapper function that you would call with your `values_of_interest` from Python, but it could be a little tricky.